src/filesystem.cpp

```diff
@@ -265,6 +277,191 @@ void FileAccess::closef()
     }
 }
 
+void FileAccess::asyncopfinished(void *param)
+{
+    AsyncIOContext *context = (AsyncIOContext *)param;
+    if (context->waiter)
+    {
+        context->waiter->notify();
+    }
+}
+
+AsyncIOContext *FileAccess::asyncfopen(string *f)
+{
+    localname.resize(1);
+    updatelocalname(f);
+
+    LOG_verbose << "Async open start";
+    AsyncIOContext *context = newasynccontext();
+    context->op = AsyncIOContext::OPEN;
+    context->access = AsyncIOContext::ACCESS_READ;
+
+    context->buffer = (byte *)f->data();
+    context->len = f->size();
+    context->waiter = waiter;
+    context->userCallback = asyncopfinished;
+    context->userData = context;
+    context->pos = size;
+    context->fa = this;
+
+    context->failed = !sysstat(&mtime, &size);
+    context->retry = this->retry;
+    context->finished = true;
+    context->userCallback(context->userData);
+    return context;
+}
+
+bool FileAccess::asyncopenf()
+{
+    if (!localname.size())
+    {
+        return true;
```

Comment:
This is the regular code path when the file is open using the fopen specific of each platform. The code below is intended to support file reads during uploads, without keeping the file open during the whole transfer.